File: ios/Sources/ForgeIOSUI/WorkflowWizardRenderer.swift
import SwiftUI
import ForgeIOSRuntime

struct WorkflowWizardRenderer: View {
    let runtime: ForgeRuntime?
    let window: WindowContext?
    let container: ContainerDef
    let form: [String: JSONValue]
    let collection: [[String: JSONValue]]
    let metrics: [String: JSONValue]
    let windowForm: [String: JSONValue]

    @State private var currentStepId = ""
    @AccessibilityFocusState private var stepFocused: Bool

    private var wizard: WizardSpec { container.wizard! }
    private var visibleSteps: [WizardStepSpec] {
        wizard.steps.filter { conditionAllows($0.visibleWhen) }
    }
    private var currentIndex: Int {
        let index = visibleSteps.firstIndex { $0.id == currentStepId } ?? 0
        return min(max(index, 0), max(visibleSteps.count - 1, 0))
    }
    private var currentStep: WizardStepSpec? { visibleSteps.indices.contains(currentIndex) ? visibleSteps[currentIndex] : nil }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(Array(visibleSteps.enumerated()), id: \.element.id) { index, step in
                        Text(step.label)
                            .font(.subheadline.weight(index == currentIndex ? .semibold : .regular))
                            .foregroundStyle(index == currentIndex ? .primary : .secondary)
                            .padding(.horizontal, 10)
                            .padding(.vertical, 6)
                            .background(index == currentIndex ? Color.accentColor.opacity(0.12) : Color.clear, in: Capsule())
                    }
                }
            }
            Group {
                if let step = currentStep,
                   let childId = step.containerId,
                   let child = container.containers.first(where: { $0.id == childId }) {
                    ContainerRenderer(
                        runtime: runtime,
                        window: window,
                        container: child,
                        inheritedDataSourceRef: container.dataSourceRef
                    )
                } else {
                    ContentUnavailableView("No active step", systemImage: "list.number")
                }
            }
            .accessibilityFocused($stepFocused)
            .accessibilityLabel(currentStep.map { "Step \(currentIndex + 1): \($0.label)" } ?? "No active step")

            HStack {
                Button("Back") { select(index: currentIndex - 1) }
                    .disabled(currentIndex <= 0)
                Spacer()
                if currentIndex < visibleSteps.count - 1 {
                    Button("Next") { select(index: currentIndex + 1) }
                        .buttonStyle(.borderedProminent)
                        .disabled(currentStep.map { !conditionAllows($0.validWhen) } ?? true)
                } else if let submit = wizard.submit,
                          let runtime,
                          let window {
                    MutationCommandButton(
                        runtime: runtime,
                        window: window,
                        sourceDataSourceRef: container.dataSourceRef ?? submit.dataSourceRef,
                        command: submit,
                        labelOverride: submit.label ?? "Submit"
                    )
                }
            }
        }
        .task(id: window?.windowID) {
            let persisted = wizard.stateKey.flatMap { windowForm[$0]?.stringValue }
            currentStepId = visibleSteps.contains { $0.id == persisted } ? (persisted ?? "") : (visibleSteps.first?.id ?? "")
        }
        .onChange(of: visibleSteps.map(\.id)) { _, ids in
            if !ids.contains(currentStepId) { select(index: 0) }
        }
        .accessibilityIdentifier("forge-wizard")
    }

    private func select(index: Int) {
        guard visibleSteps.indices.contains(index) else { return }
        currentStepId = visibleSteps[index].id
        stepFocused = true
        guard let key = wizard.stateKey, let runtime, let window else { return }
        Task { await runtime.setWindowFormValue(windowID: window.windowID, values: [key: .string(currentStepId)], bumpPrefillRevision: false) }
    }

    private func conditionAllows(_ condition: DashboardConditionDef?) -> Bool {
        DashboardRuntime.evaluateDashboardCondition(
            condition,
            metrics: metrics.mapValues(wizardAnyValue),
            form: form.mapValues(wizardAnyValue),
            windowForm: windowForm.mapValues(wizardAnyValue),
            collection: collection.map { $0.mapValues(wizardAnyValue) }
        )
    }
}

private func wizardAnyValue(_ value: JSONValue) -> Any {
    switch value {
    case .string(let value): return value
    case .number(let value): return value
    case .bool(let value): return value
    case .array(let values): return values.map(wizardAnyValue)
    case .object(let values): return values.mapValues(wizardAnyValue)
    case .null: return NSNull()
    }
}

File: android/sdk/src/main/java/com/viant/forgeandroid/ui/WorkflowWizardRenderer.kt
package com.viant.forgeandroid.ui

import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.viant.forgeandroid.runtime.ContainerDef
import com.viant.forgeandroid.runtime.DataSourceContext
import com.viant.forgeandroid.runtime.ForgeRuntime
import com.viant.forgeandroid.runtime.JsonUtil
import com.viant.forgeandroid.runtime.WindowContext
import com.viant.forgeandroid.runtime.evaluateDashboardCondition

@Composable
internal fun WorkflowWizardRenderer(
    runtime: ForgeRuntime,
    window: WindowContext,
    container: ContainerDef,
    context: DataSourceContext?,
    windowForm: Map<String, Any?>
) {
    val wizard = container.wizard ?: return
    val metadata by window.metadata.flow.collectAsState(initial = window.metadata.peek())
    val authorization = metadata?.authorizationSnapshot?.mapValues { JsonUtil.elementToAny(it.value) }.orEmpty()
    val form by context?.form?.flow?.collectAsState(initial = context.form.peek())
        ?: remember { mutableStateOf(emptyMap()) }
    val collection by context?.collection?.flow?.collectAsState(initial = context.collection.peek())
        ?: remember { mutableStateOf(emptyList()) }
    val metrics by context?.metrics?.flow?.collectAsState(initial = context.metrics.peek())
        ?: remember { mutableStateOf(emptyMap()) }
    val visibleSteps = wizard.steps.filter {
        evaluateDashboardCondition(it.visibleWhen, metrics = metrics, form = form, windowForm = windowForm, collection = collection, authorization = authorization)
    }
    val persisted = wizard.stateKey?.let { windowForm[it]?.toString() }
    var currentStepId by remember(wizard.stateKey, persisted, visibleSteps.map { it.id }) {
        mutableStateOf(persisted?.takeIf { id -> visibleSteps.any { it.id == id } } ?: visibleSteps.firstOrNull()?.id.orEmpty())
    }
    val currentIndex = visibleSteps.indexOfFirst { it.id == currentStepId }.coerceAtLeast(0)
    val current = visibleSteps.getOrNull(currentIndex)

    fun select(index: Int) {
        val step = visibleSteps.getOrNull(index) ?: return
        currentStepId = step.id
        wizard.stateKey?.let { runtime.setWindowFormValues(window.windowId, mapOf(it to step.id), bumpPrefillRevision = false) }
    }

    Column(verticalArrangement = Arrangement.spacedBy(12.dp), modifier = Modifier.semantics { contentDescription = "Wizard" }) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            modifier = Modifier.fillMaxWidth().horizontalScroll(rememberScrollState())
        ) {
            visibleSteps.forEachIndexed { index, step ->
                Surface(
                    color = if (index == currentIndex) MaterialTheme.colorScheme.primaryContainer else MaterialTheme.colorScheme.surface,
                    shape = RoundedCornerShape(999.dp)
                ) {
                    Text(step.label, modifier = Modifier.padding(horizontal = 10.dp, vertical = 6.dp))
                }
            }
        }
        current?.containerId?.let { childId ->
            container.containers.firstOrNull { it.id == childId }?.let { child ->
                ContainerRenderer(runtime, window, child, inheritedDataSourceRef = container.dataSourceRef)
            }
        }
        Row(horizontalArrangement = Arrangement.SpaceBetween, modifier = Modifier.fillMaxWidth()) {
            Button(enabled = currentIndex > 0, onClick = { select(currentIndex - 1) }) { Text("Back") }
            if (currentIndex < visibleSteps.lastIndex) {
                Button(
                    enabled = current?.let { evaluateDashboardCondition(it.validWhen, metrics = metrics, form = form, windowForm = windowForm, collection = collection, authorization = authorization) } == true,
                    onClick = { select(currentIndex + 1) }
                ) { Text("Next") }
            } else if (wizard.submit != null && context != null) {
                MutationCommandButton(
                    runtime,
                    window,
                    context,
                    wizard.submit,
                    labelOverride = wizard.submit.label ?: "Submit"
                )
            }
        }
    }
}
